File: src/lib/loader.ts
import { readFile, readdir } from 'node:fs/promises';
import { join } from 'node:path';
import { parse as parseYaml } from 'yaml';
import {
  ItemsFileSchema,
  StationsFileSchema,
  RecipesFileSchema,
} from './schema';
import type { Item, Station, Recipe } from './types';

const BIOME_ORDER: string[] = ['meadows', 'black-forest', 'swamp', 'mountain', 'plains', 'mistlands', 'ashlands'];

function computeRecipeBiome(
  recipe: Recipe,
  itemsById: Map<string, Item>,
  stationUpgradeBiomes: Map<string, Map<number, string>>,
): string | undefined {
  let maxIdx = -1;

  // Check ingredient biomes
  for (const ing of recipe.ingredients ?? []) {
    const item = itemsById.get(ing.itemId);
    if (item?.biome) {
      const idx = BIOME_ORDER.indexOf(item.biome);
      if (idx > maxIdx) maxIdx = idx;
    }
  }

  // Check station upgrade biome requirement
  const upgradeBiomes = stationUpgradeBiomes.get(recipe.station);
  if (upgradeBiomes) {
    for (let lvl = 2; lvl <= recipe.stationLevel; lvl++) {
      const biome = upgradeBiomes.get(lvl);
      if (biome) {
        const idx = BIOME_ORDER.indexOf(biome);
        if (idx > maxIdx) maxIdx = idx;
      }
    }
  }

  return maxIdx >= 0 ? BIOME_ORDER[maxIdx] : undefined;
}

export interface DataSet {
  items: Item[];
  stations: Station[];
  recipes: Recipe[];
}

async function parseYamlFile<T>(
  path: string,
  schema: { parse: (raw: unknown) => T },
): Promise<T> {
  const raw = await readFile(path, 'utf8');
  const parsed = parseYaml(raw);
  try {
    return schema.parse(parsed);
  } catch (err) {
    throw new Error(`Invalid YAML at ${path}: ${(err as Error).message}`);
  }
}

export async function loadAll(dataRoot: string): Promise<DataSet> {
  const items = await parseYamlFile(join(dataRoot, 'items.yaml'), ItemsFileSchema);
  const stations = await parseYamlFile(
    join(dataRoot, 'stations.yaml'),
    StationsFileSchema,
  );

  const recipesDir = join(dataRoot, 'recipes');
  const files = (await readdir(recipesDir)).filter((f) => f.endsWith('.yaml'));
  const recipes: Recipe[] = [];
  for (const f of files) {
    const parsed = await parseYamlFile(join(recipesDir, f), RecipesFileSchema);
    recipes.push(...parsed);
  }

  // Generate building pseudo-recipes from station upgrades
  for (const station of stations) {
    for (const upgrade of station.upgrades) {
      recipes.push({
        id: `upgrade-${station.id}-${upgrade.level}`,
        name: upgrade.name ?? `${station.name} Level ${upgrade.level}`,
        type: 'building',
        station: station.id,
        stationLevel: upgrade.level,
        ingredients: upgrade.requires,
        tags: ['station-upgrade'],
      });
    }
  }

  // Compute station upgrade biomes
  const itemsById = new Map(items.map((i) => [i.id, i]));
  const stationUpgradeBiomes = new Map<string, Map<number, string>>();
  for (const station of stations) {
    const levelBiomes = new Map<number, string>();
    for (const upgrade of station.upgrades) {
      let maxIdx = -1;
      for (const req of upgrade.requires) {
        const item = itemsById.get(req.itemId);
        if (item?.biome) {
          const idx = BIOME_ORDER.indexOf(item.biome);
          if (idx > maxIdx) maxIdx = idx;
        }
      }
      if (maxIdx >= 0) levelBiomes.set(upgrade.level, BIOME_ORDER[maxIdx]);
    }
    stationUpgradeBiomes.set(station.id, levelBiomes);
  }

  // Assign computed biome to each recipe
  for (const recipe of recipes) {
    const biome = computeRecipeBiome(recipe, itemsById, stationUpgradeBiomes);
    if (biome) (recipe as any).biome = biome;
  }

  const data: DataSet = { items, stations, recipes };
  validateCrossReferences(data);
  return data;
}

export function validateCrossReferences(data: DataSet): void {
  const itemIds = new Set(data.items.map((i) => i.id));
  const stationsById = new Map(data.stations.map((s) => [s.id, s]));
  const recipeIds = new Set<string>();

  for (const r of data.recipes) {
    if (recipeIds.has(r.id)) {
      throw new Error(`Duplicate recipe id: ${r.id}`);
    }
    recipeIds.add(r.id);

    const station = stationsById.get(r.station);
    if (!station) {
      throw new Error(
        `Recipe "${r.id}" references unknown station: ${r.station}`,
      );
    }
    if (r.stationLevel > station.maxLevel) {
      throw new Error(
        `Recipe "${r.id}" stationLevel ${r.stationLevel} exceeds ${station.id}.maxLevel ${station.maxLevel}`,
      );
    }

    for (const ing of r.ingredients ?? []) {
      if (!itemIds.has(ing.itemId)) {
        throw new Error(
          `Recipe "${r.id}" references unknown item id: ${ing.itemId}`,
        );
      }
    }

    if (r.yields && !itemIds.has(r.yields.itemId)) {
      throw new Error(
        `Recipe "${r.id}" yields unknown item id: ${r.yields.itemId}`,
      );
    }
  }
}

File: crates/valheim-data/src/loader.rs
use std::collections::HashSet;
use std::path::Path;

use thiserror::Error;

use crate::model::{GameData, Item, Recipe, RecipeType, Station};

// ── Error ──────────────────────────────────────────────────────────────────────

#[derive(Debug, Error)]
pub enum LoadError {
    #[error("failed to read {path}: {source}")]
    Io { path: String, source: std::io::Error },
    #[error("failed to parse {path}: {source}")]
    Yaml { path: String, source: serde_yaml::Error },
    #[error("validation errors:\n{0}")]
    Validation(String),
}

// ── Helpers ────────────────────────────────────────────────────────────────────

fn read_yaml<T: serde::de::DeserializeOwned>(path: &Path) -> Result<T, LoadError> {
    let contents = std::fs::read_to_string(path).map_err(|e| LoadError::Io {
        path: path.display().to_string(),
        source: e,
    })?;
    serde_yaml::from_str(&contents).map_err(|e| LoadError::Yaml {
        path: path.display().to_string(),
        source: e,
    })
}

// ── Public loader functions ────────────────────────────────────────────────────

pub fn load_items(data_dir: &Path) -> Result<Vec<Item>, LoadError> {
    read_yaml(&data_dir.join("items.yaml"))
}

pub fn load_stations(data_dir: &Path) -> Result<Vec<Station>, LoadError> {
    read_yaml(&data_dir.join("stations.yaml"))
}

pub fn load_recipes(data_dir: &Path) -> Result<Vec<Recipe>, LoadError> {
    let recipes_dir = data_dir.join("recipes");
    let mut crafting: Vec<Recipe> = read_yaml(&recipes_dir.join("crafting.yaml"))?;
    let cooking: Vec<Recipe> = read_yaml(&recipes_dir.join("cooking.yaml"))?;
    crafting.extend(cooking);
    Ok(crafting)
}

// ── Building recipe generation ─────────────────────────────────────────────────

pub fn generate_building_recipes(stations: &[Station]) -> Vec<Recipe> {
    let mut recipes = Vec::new();
    for station in stations {
        for upgrade in &station.upgrades {
            let level = upgrade.level;
            let id = format!("upgrade-{}-{}", station.id, level);
            let name = upgrade
                .name
                .clone()
                .unwrap_or_else(|| format!("{} Level {}", station.name, level));
            let recipe = Recipe {
                id,
                name,
                recipe_type: RecipeType::Building,
                station: station.id.clone(),
                station_level: level.saturating_sub(1),
                repair_level: None,
                ingredients: upgrade.requires.clone(),
                yields: None,
                skill: None,
                tags: vec!["station-upgrade".to_string()],
                notes: None,
                food: None,
                secondary_step: None,
                stats: None,
                upgrades: vec![],
                biome: None,
            };
            recipes.push(recipe);
        }
    }
    recipes
}

// ── Cross-reference validation ─────────────────────────────────────────────────

pub fn validate_cross_references(
    items: &[Item],
    stations: &[Station],
    recipes: &[Recipe],
) -> Result<(), LoadError> {
    let item_ids: HashSet<&str> = items.iter().map(|i| i.id.as_str()).collect();
    let station_map: std::collections::HashMap<&str, u32> = stations
        .iter()
        .map(|s| (s.id.as_str(), s.max_level))
        .collect();

    let mut errors: Vec<String> = Vec::new();

    for recipe in recipes {
        // Check 1: recipe.station must exist (skip for Building recipes)
        if recipe.recipe_type != RecipeType::Building {
            if !station_map.contains_key(recipe.station.as_str()) {
                errors.push(format!(
                    "recipe '{}': unknown station '{}'",
                    recipe.id, recipe.station
                ));
            } else {
                // Check 2: station_level <= station.max_level
                let max = station_map[recipe.station.as_str()];
                if recipe.station_level > max {
                    errors.push(format!(
                        "recipe '{}': station_level {} exceeds max_level {} for station '{}'",
                        recipe.id, recipe.station_level, max, recipe.station
                    ));
                }
            }
        }

        // Check 3: ingredient item_ids must exist
        for ing in &recipe.ingredients {
            if !item_ids.contains(ing.item_id.as_str()) {
                errors.push(format!(
                    "recipe '{}': unknown ingredient item_id '{}'",
                    recipe.id, ing.item_id
                ));
            }
        }

        // Check 4: yields item_id must exist
        if let Some(yields) = &recipe.yields {
            if !item_ids.contains(yields.item_id.as_str()) {
                errors.push(format!(
                    "recipe '{}': unknown yields item_id '{}'",
                    recipe.id, yields.item_id
                ));
            }
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(LoadError::Validation(errors.join("\n")))
    }
}

// ── load_all ───────────────────────────────────────────────────────────────────

pub fn load_all(data_dir: &Path) -> Result<GameData, LoadError> {
    let items = load_items(data_dir)?;
    let stations = load_stations(data_dir)?;
    let mut recipes = load_recipes(data_dir)?;
    let building_recipes = generate_building_recipes(&stations);
    recipes.extend(building_recipes);
    validate_cross_references(&items, &stations, &recipes)?;
    Ok(GameData {
        items,
        stations,
        recipes,
    })
}

// ── Tests ──────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn data_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../src/data")
    }

    #[test]
    fn load_items_from_yaml() {
        let items = load_items(&data_dir()).expect("should load items");
        assert!(!items.is_empty(), "items should be non-empty");
        assert!(
            items.iter().any(|i| i.id == "iron"),
            "items should contain 'iron'"
        );
    }

    #[test]
    fn load_stations_from_yaml() {
        let stations = load_stations(&data_dir()).expect("should load stations");
        let forge = stations
            .iter()
            .find(|s| s.id == "forge")
            .expect("forge station should exist");
        assert_eq!(forge.max_level, 7, "forge should have max_level 7");
    }

    #[test]
    fn load_recipes_from_yaml() {
        let recipes = load_recipes(&data_dir()).expect("should load recipes");
        assert!(
            recipes.iter().any(|r| r.recipe_type == RecipeType::Crafting),
            "should contain crafting recipes"
        );
        assert!(
            recipes.iter().any(|r| r.recipe_type == RecipeType::Cooking),
            "should contain cooking recipes"
        );
    }

    #[test]
    fn load_all_validates_cross_references() {
        load_all(&data_dir()).expect("load_all should succeed with valid cross-references");
    }

    #[test]
    fn load_all_generates_building_recipes() {
        let data = load_all(&data_dir()).expect("should load all data");
        assert!(
            data.recipes
                .iter()
                .any(|r| r.recipe_type == RecipeType::Building),
            "should contain building recipes"
        );
    }
}
